src/Polysemy/Final/Internal.hs

```diff
@@ -0,0 +1,230 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Polysemy.Final.Internal where
+
+import Data.Functor.Compose
+import Data.Maybe
+
+import Control.Concurrent.MVar
+import Control.Exception
+import Control.Monad
+import Control.Monad.Trans.Maybe
+
+import Polysemy
+import Polysemy.Internal
+import Polysemy.Internal.Union
+
+-- | This represents a function which produces
+-- an action of the final monad @m@ given:
+--
+--   * The initial effectful state at the moment the action
+--     is to be executed.
+--
+--   * An way to convert @z@ (which is typically @'Sem' r@)to @m@ by
+--     threading the effectful state through.
+--
+--   * An inspector that is able to view some value within the
+--     effectful state if the effectful state contains any values.
+--
+-- A @'Polysemy.Internal.Union.Weaving' 'Final'@ provides these components,
+-- hence the name 'ThroughWeavingToFinal'.
+type ThroughWeavingToFinal m z a =
+     forall f
+   . Functor f
+  => f ()
+  -> (forall x. f (z x) -> m (f x))
+  -> (forall x. f x -> Maybe x)
+  -> m (f a)
```

Comment:
What do you think of this? For some reason it slipped my mind `withWeavingToFinal` can't be annotated either, so I created this type which I thoroughly document, and use it in `WithWeavingToFinal`

Reply:
love it!

Reply:
Good! I'll keep this then.